### DISCUSSION TITLE
Flaw in Database class behavior

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#122] [Design & Development]
As always exercising any software aimed at data processing is subject to surprises when you apply it to the real thing.  Just hit a couple related to the Database class.  

1.   When I run save_ensemble_data with this line:  `db.save_ensemble_data(d_cleaned,'gridfs')` on a serial test job with the usarray data set it immediately exits with this exception message: `MsPASSError: The history object to be saved has a duplicate uuid`.   That message comes from database.py in this handler in save_ensemble_data:
```
        try:
            if prev_history_object_id:
                # overwrite history
                history_col.delete_one({'_id': prev_history_object_id})
                history_col.insert_one({'_id': current_uuid, 'nodesdata': history_binary})
            else:
                # new insertion
                history_col.insert_one({'_id': current_uuid, 'nodesdata': history_binary})
        except pymongo.errors.DuplicateKeyError as e:
            raise MsPASSError("The history object to be saved has a duplicate uuid", "Fatal") from e
```
That condition should not be raising an exception.  There are lots of ways multiple data objects can contain a history record with the same uuid.  The whole idea of a uuid is it is unique, so I guess whoever did this was thinking "it s unique so don't preserve it more than once".  That is wrong, however, as the purpose in history is to allow the G-tree defining the processing history to be stored in a flat container - a multimap in the implementation.  Recall that without most implementations of trees use arrays of pointers, but when designing this we quickly realized that would not work with spark because pointers would point at the wrong thing if moved so serializing such a structure would be a nightmare.   Oddly reviewing how git works to understand git reset I was reading a tutorial and realized git more or less uses the same concept to store commit history.  That is, the implementation of git stores the commit history in a flat container that can be thought of as tree branches, but it isn't really implemented that way.  I digress, but it is an important reminder.  The point is that we aren't saving history correctly.  That should not throw an exception.  Do we also know if a reader can properly reconstruct a ProcessingHistory object from data stored in the history collection in MongoDB?

2. *save_ensemble_data* is missing an important option:   I have no way to disable saving history.  I think the current logic just keys is testing if history is or is not empty.   All writers should have the option of not saving history.  ErrorLogger (elog) data should also be optional, BUT the default should be to save all elog entries.   The only reason to ever turn it off is if some process is generating excessive verbose logging messages the user might want to handle a different way.  Rare, but something we should allow.  I think a while back we decided history should be optional and the writers should reflect that and not save history unless requested.

Note this isn't a critical fix as I have a simple, if not elegant fix.  I'll just clear history before calling save.

**Comment by pavlis** [comment]
Correction:  clear_history is not a workaround.  I am not sure why it is not.

**Comment by pavlis** [comment]
I know why.   Here is the relevant code in the _save_history method of Database:
```
        if not collection:
          collection = self.database_schema.default_name('history_object')
        history_col = self[collection]
        proc_history = ProcessingHistory(mspass_object)
        current_uuid = proc_history.id() # uuid in the current node
        history_binary = pickle.dumps(proc_history)
```

As a minimum it need this fix:
```
        proc_history = ProcessingHistory(mspass_object)
        if proc_history.is_empty():
            return
        if not collection:
            collection = self.database_schema.default_name('history_object')
        history_col = self[collection]
        current_uuid = proc_history.id() # uuid in the current node
        history_binary = pickle.dumps(proc_history)
```
I'm going to patch my local copy to validate this workflow and get some timing benchmarks.  (I need to know how feasible this workflow is to run serial on my local Mac to know if we can assemble the usarray data on that box.  That will be the fastest way right now to produce a working data set.

**Comment by pavlis** [comment]
And another I'm working on.   Once I got the history workaround I found the save_data method was not handling dead data properly.   The really bad thing it does is abort when something inside is incorrectly trying to access the data matrix with data marked dead.   That is just a symptom.  I'm going to patch the version I have to behave as I think we concluded it should:  dead data should only have elog data saved to the database.  Quick look suggests maybe we need to save the Metadata or the elog entries won't have a context - pretty sure of that but will confirm after working with a bit more.  We can discuss this some more in our conference call later today.

**Comment by pavlis** [comment]
A hack solution was easy.   The right solution is a rewrite of save_data.   Let's discuss how to make that happen in todays zoom meeting.

**Reply by pavlis** [reply]
For the record this is what I actually implemented:
```
         if not isinstance(mspass_object, (TimeSeries, Seismogram)):
            raise TypeError("only TimeSeries and Seismogram are supported")
        # This is a temporary patch to handle a known problem - dead data.
        # A proper fix needs to reconsider the entire sequence in 
        # this method. 
        if mspass_object.live:
            ok_to_save=True
        else:
            ok_to_save=False
        if storage_mode not in ['file', 'gridfs']:
            raise TypeError("Unknown storage mode: {}".format(storage_mode))
        # below we try to capture permission issue before writing anything to the database.
        # However, in the case that a storage is almost full, exceptions can still be 
        # thrown, which could mess up the database record.
        if storage_mode == 'file':
            if not dfile and not dir:
                # Note the following uses the dir and dfile defined in the data object.
                # It will ignore these two keys already in the collection in an update
                # transaction, and the dir and dfile in the collection will be replaced.
                if ('dir' not in mspass_object) or ('dfile' not in mspass_object):
                    raise ValueError(
                        'dir or dfile is not specified in data object')
                dir = os.path.abspath(mspass_object['dir'])
                dfile = mspass_object['dfile']
            else:
                dir = os.path.abspath(dir)
            fname = os.path.join(dir, dfile)
            if os.path.exists(fname):
                if not os.access(fname, os.W_OK):
                    raise PermissionError(
                        'No write permission to the save file: {}'.format(fname))
            else:
                # the following loop finds the top level of existing parents to fname
                # and check for write permission to that directory. 
                for path_item in pathlib.PurePath(fname).parents:
                    if os.path.exists(path_item):
                        if not os.access(path_item, os.W_OK | os.X_OK):
                            raise PermissionError(
                                'No write permission to the save directory: {}'.format(dir))
                        break

        schema = self.metadata_schema
        if isinstance(mspass_object, TimeSeries):
            save_schema = schema.TimeSeries
        else:
            save_schema = schema.Seismogram
        # This function is needed to make sure the metadata define the time
        # standard consistently
        self._sync_time_metadata(mspass_object)
        # 1. save metadata
        self.update_metadata(mspass_object, update_all, exclude, collection)

        # 2. save data
        # This is an ugly patch to handle dead data.  As noted 
        # above needs a cleaner solution
        if not ok_to_save:
            return
        wf_collection = save_schema.collection('_id') if not collection else collection
        col = self[wf_collection]
        object_doc = col.find_one({'_id': mspass_object['_id']})
        filter_ = {'_id': mspass_object['_id']}
        update_dict = {'storage_mode': storage_mode}

        if storage_mode == "file":
            foff = self._save_data_to_dfile(mspass_object, dir, dfile)
            update_dict['dir'] = dir
            update_dict['dfile'] = dfile
            update_dict['foff'] = foff
        elif storage_mode == "gridfs":
            old_gridfs_id = None if 'gridfs_id' not in object_doc else object_doc['gridfs_id']
            gridfs_id = self._save_data_to_gridfs(mspass_object, old_gridfs_id)
            update_dict['gridfs_id'] = gridfs_id
        #TODO will support url mode later 
        #elif storage_mode == "url":
        #    pass

        col.update_one(filter_, {'$set': update_dict})
```
I don't like this structure because:  (1) it has to use the (unnecessary if the code order is changed) bool "ok_to_save" and (2) the conditional with a return is kind of bad form in a python program.   If you rework this, as I think we decided you would, I would reconsider the order of operations of the set of things done here.  The code needs to handle dead data more elegantly.

**Comment by wangyinz** [comment]
Hmmm.... I am confused. By design, the history's uuid shouldn't be the same because that's the index. I guess you got the same uuid with a copy operation? Maybe we should consider changing the behavior of any copy such that it always generate a new uuid. Current implementation serializes the whole ProcessingHistory object so there is no tree structure in the collection but only a one-on-one mapping between the wf entries and the history_object entries. So I don't think it make sense to have duplicated uuid in this case.

Yeah, the `save_data` currently assumes the data is clean, which may not be true. We definitely need to add the logic to handle dead object there. However, I don't know how to preserve the elog with current schema design. The elog entries are also one-on-one mapped to the wf entries. I guess we have to first finish our global history related schema and have elog entries linked to a global history record to avoid a dangling record in the elog, and actually that's why I tend to not save elogs of dead objects in this version and revise it later in our global history.

**Comment by pavlis** [comment]
Let me close this discussion with a summary of what we concluded in our meeting earlier today:
1.  I looked at the code for saving history and realize I was wrong about how uuids were handled in that situation.  The errors I was getting were purely from a programming bug upstream and not in the section saving history.   
2. Item 1 is true as long as we do what the current code does with uses the uuid at the root of the inverted history tree = the uuid of the object in it's current state.
3. There was a bug in the section of Database that was not handling the situation correctly when history was empty.

**Comment by wangyinz** [comment]
Issues discussed here should be addressed by #125
